elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/metadata/models/Metric.java

```diff
@@ -45,27 +53,146 @@ public Metric(Table table, String fieldName, EntityDictionary dictionary) {
+                    MetricFormula.class,
+                    fieldName);
+
+            if (formula != null) {
+                this.metricFunction = constructMetricFunction(
+                        constructColumnName(tableClass, fieldName, dictionary) + "[" + fieldName + "]",
+                        meta == null ? null : meta.longName(),
+                        meta == null ? null : meta.description(),
+                        formula.value(),
+                        new HashSet<>());
+
+            } else {
+                throw new IllegalArgumentException("Trying to construct metric field "
+                        + getId() + " without @MetricAggregation and @MetricFormula.");
+            }
+        }
+    }
+
+    /**
+     * Resolve aggregation function from {@link MetricAggregation} annotation.
+     *
+     * @param tableClass table class
+     * @param fieldName metric field name
+     * @param aggregation aggregation annotation on the field
+     * @param meta meta annotation on the field
+     * @param dictionary dictionary with entity information
+     * @return resolved metric function instance
+     */
+    private static MetricFunction resolveAggregation(Class<?> tableClass,
+                                                     String fieldName,
+                                                     MetricAggregation aggregation,
+                                                     Meta meta,
+                                                     EntityDictionary dictionary) {
+        String columnName = constructColumnName(tableClass, fieldName, dictionary);
         try {
-            this.metricFunction = metric.function().newInstance();
-            metricFunction.setName(getId() + "[" + metricFunction.getName() + "]");
-            metricFunction.setExpression(String.format(
-                    metricFunction.getExpression(),
-                    dictionary.getAnnotatedColumnName(tableClass, fieldName)));
+            MetricFunction metricFunction = aggregation.function().newInstance();
+            metricFunction.setName(columnName + "[" + metricFunction.getName() + "]");
 
             if (meta != null) {
                 metricFunction.setLongName(meta.longName());
                 metricFunction.setDescription(meta.description());
             }
+
+            return metricFunction;
         } catch (InstantiationException | IllegalAccessException e) {
-            throw new IllegalArgumentException("Can't initialize function for metric " + getId());
+            throw new IllegalArgumentException("Can't initialize function for metric " + columnName);
         }
     }
 
+    /**
+     * Dynamically construct a metric function
+     *
+     * @param id metric function id
+     * @param longName meta long name
+     * @param description meta description
+     * @param expression expression string
+     * @param arguments function arguments
+     * @return a metric function instance
+     */
```

Comment:
I think it is clearer to just call the constructor directly.  I would remove this function.